src/sim/TrackInitUtils.hh

```diff
@@ -16,13 +19,43 @@ namespace celeritas
 //---------------------------------------------------------------------------//
 // HELPER FUNCTIONS
 //---------------------------------------------------------------------------//
-// Create track initializers on device from primary particles
-void extend_from_primaries(const TrackInitParamsHostRef& params,
-                           TrackInitStateDeviceVal*      data);
+/*!
+ * Create track initializers from primary particles.
+ *
+ * This creates the maximum possible number of track initializers on the
+ * templated memory space from host primaries (either the number of host
+ * primaries that have not yet been initialized or the size of the available
+ * storage in the track initializer vector, whichever is smaller).
+ */
+template<MemSpace M>
+inline void extend_from_primaries(const TrackInitParamsHostRef& params,
+                                  TrackInitStateData<Ownership::value, M>* data)
+{
+    CELER_EXPECT(params);
+    CELER_EXPECT(data && *data);
 
-// Create track initializers on host from primary particles
-void extend_from_primaries(const TrackInitParamsHostRef& params,
-                           TrackInitStateHostVal*        data);
+    // Number of primaries to initialize
+    auto count = min(data->initializers.capacity() - data->initializers.size(),
+                     data->num_primaries);
+    if (count > 0)
+    {
+        data->initializers.resize(data->initializers.size() + count);
+
+        // Allocate memory and copy primaries
+        Collection<Primary, Ownership::value, M> primaries;
+        make_builder(&primaries).resize(count);
+        detail::Copier<Primary, MemSpace::host> copy{
```

Comment:
`detail` functions/classes are intended to be used only by some file in the parent directory (`base/Collection` in this case). If this class is useful enough to be needed elsewhere we should expose it in the main namespace and add a test for it.

Reply:
Ah, right -- I'll move it out of `detail` and add a test. Thanks!